### DISCUSSION TITLE
How to **read** __user variable in kprobe trace or **copy** it to a bpftrace variable

### DISCUSSION THREAD

**Discussion by mebuis** [bpftrace/bpftrace#2603] [Q&A]
Please forgive my english.

I'm working in the Ubunt22.04 with kernel 5.15
I need to trace a kernel function in syscall table with kprobe like _sys_syslog_.
it define's asmlinkage _long sys_syslog(int type, char __user *buffer, int len);_

I hope I can **read** the variable buffer in kprobe trace with bpftrace. Or **copy** it to a bpftrace variable($a, @b etc.)

How do I use the current bpftrace implementation?

I'm tried use function `buf` like `printf("buffer:%r\n",buf(uptr(buffer))` and run with command `bpftrace -kk my.bt`.  It running but not working!!

There are many functions that use the `__user` modifier in syscall table. How to solve it??

**Comment by viktormalik** [comment]
Hi @mebuis, this seems to be more a question than an issue, so I took the liberty to convert it into a discussion.

For tracing syscalls, I'd generally recommend using the `tracepoint:syscalls:sys_enter_*` probes. The arguments of the syscalls are then available via `args->`.

For your case, to e.g. print the buffer, that would be:
```
# bpftrace -e 'tracepoint:syscall:sys_enter_syslog { printf("%s", str(args->buf); }'
```
To get the complete list of syscall arguments, you can use
```
# bpftrace -lv 'tracepoint:syscalls:sys_enter_syslog
tracepoint:syscalls:sys_enter_syslog
    int __syscall_nr
    int type
    char * buf
    int len
```

**Reply by mebuis** [reply]
Thansk for your kind reply!

But,if the arg is not a `char*`, it's a struct, can bpftrace copy the struct to kernel?
I need use this function to process more logic.

and, I have some other questions
1. how to get buffer from bpftrace function `buf` like `$code = (int8)buf(kaddr("sys_open"),1);`
2. how to get file content to bpftrace varialbe like `$content = cat("/etc/fstab");`
3. how to traversing a tuple like

```
$funcs = ("__x64_sys_ioctl","__x64_sys_ioperm","__x64_sys_keyctl","__x64_sys_landlock_add_rule","__x64_sys_landlock_create_ruleset");

$i = 0;
while($i < 5){
	printf("funcName:%s\n", $funcs.$i);
	++$i;
}
```
I need these functions for security research.
Now,I thought the bpftrace is good at performance tuning,but not very good at security research.

**Reply by viktormalik** [reply]
Is it a struct or a struct pointer? I'm not sure if we can handle struct arguments, yet. If it's a pointer, then it's not a problem, you can store the pointed struct in a local variable, map, print it, etc.

As for the questions:
1. I don't understand what you're trying to achieve here. `kaddr` returns address of the given symbol. `sys_open` doesn't even exist but if you mean the syscall handler, it's called `__x86_sys_open` on x86. It is a function, so its address is the address of the first instruction. Using `buf(kaddr("__x86_sys_open"), 1)` will give you hexadecimal representation of the first byte of the first instruction.
2. This is not possible. We have the `system` builtin call but you cannot assign its stdout into a variable.
3. Also not possible as we cannot iterate tuples. But if you described more concretely what you're trying to achieve, maybe we can suggest some feasible solution.

**Reply by mebuis** [reply]
Hi @viktormalik , thanks for your kind reply and Pls forgive my bad English.
Now,I'm tring to write a  check script of inline hook with bpftrace.

The basic principle is to check the function header instruction.

I wrote a bpftrace one-liner like this:
`bpftrace -e 'BEGIN{printf("%r\n",buf(kaddr("__x64_sys_kill"),16));exit()}'`

It works well and output the following:
```
Attaching 1 probe...
\xe8\x8b\x1bG5U\xb9\x06\x00\x00\x00\xbe\x01\x00\x00\x00
```
The first instruction is 0xE8，So I think the function is hooked.

But I need the script to automatically judge, if the first instruction is 0xE8 or 0xE9, the script will print msg to tell me which kernel functions are hooked instead of making artificial judgments.
The pseudocode is roughly as follows:
```
$code = (unsigned char)buf(kaddr(@funcname[$i]),1);
if($code == 0xe9 || $code == 0xe8){
	printf("function %s is hooked\n", @funcname[$i]);
	delete(@funcname[$i]);
}
```

To go further, I have several functions to check,I need put these functions in a file or a tuple.And then run batch tests to determine if the system has been hacked.

I really don't know how to do that. Right. It's the first time to write a script with bpftrace for me.
